XLS-0066d-lending-protocol/README.md

```diff
@@ -0,0 +1,1428 @@
+$$
+
+The `principal` and `interest` portions can be derived as follows:
+
+$$
+interest = principalOutstanding \times periodicRate
+$$
+
+$$
+principal = periodicPayment - interest
+$$
+
+###### 3.2.5.1.2 Late Payment
+
+When a Borrower makes a payment after `NextPaymentDueDate`, they must pay a nominal late payment fee and an additional interest rate charged on the overdue amount for the unpaid period. The formula is as follows:
+
+$$
+totalDue = periodicPayment + latePaymentFee + latePaymentInterest
+$$
+
+A special, late payment interest rate is applied for the over-due period:
+
+$$
+latePaymentInterest = principalOutstanding \times \frac{lateInterestRate \times secondsSinceLastPayment}{365 \times 24 \times 60 \times 60}
+$$
+
+A late payment pays more interest than calculated when increasing the Vault value in the `LoanSet` transaction. Therefore, the total Vault value captured by `Vault.AssetTotal` must be recalculate
+d.
+Assume the function `PeriodicPayment()` returns the expected periodic payment, split into `principalPeriodic` and `interestPeriodic`. Furthermore, assume the function `LatePayment()` that implements the Late Payment formula. The function returns the late payment split into `principalLate` and `interestLate`, where `interestLate` is calculated using the formula above. Note that `principalPeriodic == principalLate` and `interestLate > interestPeriodic` are used only when the payment is late. Otherwise, `interestLate == interestPeriodic`.
+
+$$
+valueChange = interestLate - interestPeriodic
+$$
+
+Note that `valueChange >= 0`.
+
+###### 3.2.5.1.3 Loan Overpayment
+
+- Let $\mathcal{P}$ and $\mathcal{p}$ represent the total and outstanding Loan principal.
+- Let $\mathcal{I}$ and $\mathcal{i}$ represent the total and outstanding Loan interest computed from $\mathcal{P}$ and $\mathcal{p}$ respectively.
+
+$$
+excess = min(\mathcal{p}, paymentAmountMade - minimumPaymentAmount)
+$$
+
+$$
+interestPortion = excess \times overpaymentInterestRate
+$$
+
+$$
+feePortion = excess \times overpaymentFee
+$$
+
+$$
+principalPortion = excess - interestPortion - feePortion
+$$
+
+$$
+\mathcal{p'} = \mathcal{p} - principalPortion
+$$
+
+Let $\mathcal{i}$ denote the outstanding interest computed from $\mathcal{p}$. Simillarly, let $\mathcal{i'}$ denote the outstanding interest computed from $\mathcal{p'}$. We compute the loan interest change as follows:
+
+$$
+valueChange =  \mathcal{i} - \mathcal{i'}
+$$
+
+###### 3.2.5.1.4 Early Full Repayment
+
+A Borrower can close a Loan early by submitting the total amount needed to do so. This amount is the sum of the remaining balance, any accrued interest, a prepayment penalty, and a prepayment fee.
+
+$$
+totalDue = principalOutstanding + accruedInterest + prepaymentPenalty + ClosePaymentFee
+$$
+
+Accrued interest up to the point of early closure is calculated as follows:
+
+$$
+accruedInterest = principalOutstanding  \times periodicRate \times \frac{secondsSinceLastPayment}{paymentInterval}
+$$
```

Comment:
Just to clarify, the only way to reliably compute the value of `secondsSinceLastPayment` is `lastLedgerCloseTime - Loan.previousPaymentDate`. That will mean with a 3-4 second ledger close time, the lender is going to miss out on 3-4 seconds of interest. I hope they can survive the shortfall. 😈

Reply:
Yeah, this is the downside of a second-based interest calculation, that in essence the time "jumps" 3 -5 seconds. That said, doing 5 second interest accumulation would also be confusing, so I guess they'll have to accept the shortfall :thinking:

Reply:
> Yeah, this is the downside of a second-based interest calculation, that in essence the time "jumps" 3 -5 seconds. That said, doing 5 second interest accumulation would also be confusing, so I guess they'll have to accept the shortfall 🤔

Sounds good!